feat: additional rule parsing, core_match updates for validators

kopecs opened the pull request:
Modifications to parsing and core_match to enable:
- validators to be able to rewrite some fields in cli_matches
- matching on HTTP response content

Test plan: see corresponding pro PR.

kopecs committed:
Implement output modification in core

Adds required parsing code and fleshes out rule/severity to account for
these changes. Enables modification of output fields based on which
validator matches in a secrets rule.

This also moves severity to Pattern_match from Rule to avoid a circular
dependency between the two.

kopecs committed:
Handle cli_match updates in the CLI

For severity, message and metadata, use extras value for override if
present, otherwise use rule info. Note that metadata is not a straight
replacment, but rather we merge the dictionaries, preferring the
override version in the case of a key present in both.

kopecs committed:
Get submodule update for core match

aryx commented on cli/src/semgrep/core_output.py:
```diff
@@ -19,6 +19,7 @@
 
 import semgrep.semgrep_interfaces.semgrep_output_v1 as out
 import semgrep.util as util
+from semgrep.constants import RuleSeverity
```
Let's reuse the (new) severity in semgrep_output_v1.atd instead

aryx replied:
I'll do that in a separate PR

aryx commented on src/core/Mini_rule.ml:
```diff
@@ -37,7 +37,7 @@ type rule = {
   (* originates from pattern-inside or pattern-not-inside *)
   message : string;
-  severity : Rule.severity;
+  severity : Pattern_match.severity;
```
Let's reuse the (new) severity in semgrep_output_v1.atd

aryx commented on src/core/Pattern_match.ml:
```diff
@@ -77,6 +77,13 @@ type taint_trace_item = {
 
 type taint_trace = taint_trace_item list [@@deriving show, eq]
 
+(* This type indicates the severity of a finding. Typically this is
```
move this comment in semgrep_output_v1.atd now and define the type there.

aryx commented on src/core/Rule.ml:
```diff
@@ -371,15 +371,17 @@ type http_match_clause = {
   (* Optional. Empty list if not set *)
   headers : header list;
-  content : JSON.t option; (* TODO(cooper): use contents for matching body *)
+  content : (formula * Xlang.t) option;
```
content is not a great name. Should it be body_formula instead?

aryx reviewed (changes requested):
LGTM but let's fix first this severity thing.

kopecs committed:
severity wip (todo rebase on pad)

kopecs committed:
Work toward unified severity type

Still requires work to integrate with the polymorphic variant version
from Severity.ml

kopecs committed:
fixup! Work toward unified severity type

kopecs committed:
Make o and pysemgrep metadata update consistent

Changes the ocaml function implementing this logic for consistency.

aryx committed:
Merge branch 'develop' into cooper/validator-syntax-p2

aryx commented on src/osemgrep/cli_ci/Ci_subcommand.ml:
```diff
@@ -302,10 +302,11 @@ let partition_findings ~keep_ignored (results : Out.cli_match list) =
+  | Experiment -> `Int 4
+  | Warning -> `Int 1
+  | Error -> `Int 2
   | _ -> `Int 0
```
better be explicit and avoid _ (actually we have a semgrep rule against that; not sure why it didn't triggered here).

aryx commented on src/core/Pattern_match.ml:
```diff
@@ -91,6 +91,8 @@ type validation_state =
   | No_validator
 [@@deriving show, eq]
 
+type severity = Semgrep_output_v1_t.severity [@@deriving show, eq]
```
not sure it's worth introducing an alias. I'll fix that in a separate PR.

aryx commented on src/osemgrep/cli_ci/Ci_subcommand.ml:
```diff
@@ -446,7 +447,8 @@ let findings_and_complete ~has_blocking_findings ~commit_date ~engine_requested
     new_ignored
-    |> List.exists (fun m -> String.equal m.Out.extra.severity "EXPERIMENT")
+    |> List.exists (fun (m : Out.cli_match) ->
+           [%eq: Out.severity] m.extra.severity Experiment)
```
I use =*= for those

aryx commented on src/osemgrep/cli_scan/Output.ml:
```diff
@@ -100,7 +100,7 @@ let dispatch_output_format (output_format : Output_format.t)
               extra = { message; severity; _ };
               _;
              } ->
-                 let severity = String.lowercase_ascii severity in
```
seems like a different change here ... we use to lowercase here.

aryx committed:
fix mypy and CI

aryx committed:
fix osemgrep regressions

aryx commented:
The semgrep-pro crash is normal, it's because the core_output format changed
